pss/pss_test.go

```diff
@@ -507,116 +510,128 @@ func TestMessageProcessing(t *testing.T) {
-	default:
+	usedSlots := ps.outbox.len()
+	if usedSlots != 1 {
+		t.Fatal("Outbox len should be 1 before finish processing")
 	}
-	if outmsg != nil {
-		t.Fatalf("expected outbox empty after full address on msg, but had message %s", msg)
+	t.Log("Message enqueued", "Outbox len", ps.outbox.len())
+
+	<-successC
+	if len(processed) != 1 {
+		t.Fatal("There should be one message processed")
 	}
 
-	// message should pass and queue due to partial length
-	msg.To = addr[0:1]
-	msg.Payload.Data = []byte{0x78, 0x79, 0x80, 0x80, 0x79}
-	if err := ps.handle(context.TODO(), msg); err != nil {
-		t.Fatal(err.Error())
+	failed := make([]*PssMsg, 0)
+	failedC := make(chan struct{})
+	continueC := make(chan struct{})
+	failedForward := func(msg *PssMsg) bool {
+		failed = append(processed, msg)
+		failedC <- struct{}{}
+		<-continueC
+		return false
 	}
-	tmr.Reset(time.Millisecond * 100)
-	outmsg = nil
-	select {
-	case outmsg = <-ps.outbox:
-	case <-tmr.C:
+
+	ps.forwardFunc = failedForward
+
+	err = ps.enqueue(testRandomMessage())
+	if err != nil {
+		t.Fatal("Message should have space for enqueuing")
 	}
-	if outmsg == nil {
-		t.Fatal("expected message in outbox on encrypt fail, but empty")
+
+	<-failedC
+	if len(failed) == 0 {
+		t.Fatal("There should be one failed messages")
 	}
-	outmsg = nil
-	select {
-	case outmsg = <-ps.outbox:
-	default:
+	if len(processed) > 1 {
+		t.Fatal("There should be only one message processed")
 	}
-	if outmsg != nil {
-		t.Fatalf("expected only one queued message but also had message %v", msg)
+	// The message will be retried once we send to continueC, so first, we change the forward function
+	ps.forwardFunc = forward
+	continueC <- struct{}{}
+	<-successC
+	if len(processed) != 2 {
+		t.Fatal("There should be two message processed", "len(processed)", len(processed))
 	}
 
-	// full address mismatch should put message in queue
-	msg.To[0] = 0xff
-	if err := ps.handle(context.TODO(), msg); err != nil {
+}
+
+func TestOutboxFull(t *testing.T) {
+	// setup
+	privkey, err := crypto.GenerateKey()
+	if err != nil {
 		t.Fatal(err.Error())
 	}
-	tmr.Reset(time.Millisecond * 10)
-	outmsg = nil
-	select {
-	case outmsg = <-ps.outbox:
-	case <-tmr.C:
-	}
-	if outmsg == nil {
-		t.Fatal("expected message in outbox on address mismatch, but empty")
-	}
-	outmsg = nil
-	select {
-	case outmsg = <-ps.outbox:
-	default:
-	}
-	if outmsg != nil {
-		t.Fatalf("expected only one queued message but also had message %v", msg)
+
+	addr := make([]byte, 32)
+	addr[0] = 0x01
+	ps := newTestPssStart(privkey, network.NewKademlia(addr, network.NewKadParams()), NewParams(), false)
+	defer ps.Stop()
+	outboxCapacity := 2
+
+	processed := make([]*PssMsg, 0)
+	procChan := make(chan struct{})
+	ps.outbox = newOutbox(outboxCapacity)
+	succesForward := func(msg *PssMsg) bool {
+		log.Info("Processing message")
+		<-procChan
+		log.Info("Message processed")
+		processed = append(processed, msg)
+		return true
 	}
+	ps.forwardFunc = succesForward
+
```

Comment:
it seems the actual test part is missing here no?
i.e., check

Reply:
You are right, must have lost it at some point

Reply:
Implemented in 1b91b66